File: go/sorting/merge_sort.go
package sorting

type MergeSort struct {
	workArr []int
}

func (m MergeSort) Sort(arr []int, sortType SortType) []int {
	length := len(arr)
	m.workArr = make([]int, length)
	return m.MergeSort(arr, 0, length-1, sortType)
}

func (m MergeSort) MergeSort(arr []int, low int, high int, sortType SortType) []int {
	if low >= high {
		return arr
	}

	middle := (low + high) / 2

	m.MergeSort(arr, low, middle, sortType)
	m.MergeSort(arr, middle+1, high, sortType)

	for i := low; i <= middle; i++ {
		m.workArr[i] = arr[i]
	}

	j := high
	for i := middle + 1; i <= high; i++ {
		m.workArr[i] = arr[j]
		j--
	}

	i := low
	j = high

	for k := low; k <= high; k++ {
		if Compare(m.workArr[j], m.workArr[i], sortType) {
			arr[k] = m.workArr[i]
			i++
		} else {
			arr[k] = m.workArr[j]
			j--
		}
	}

	return arr
}

File: dart/lib/sorting/merge_sort.dart
import 'package:algorithm/sorting/sort.dart';

class MergeSort extends Sort {
  List<int> workArr;

  @override
  List<int> sort(List<int> arr, {SortType type = SortType.asc}) {
    final length = arr.length;
    workArr = []..length = length;
    return mergeSort(arr, 0, length - 1, type);
  }

  List<int> mergeSort(List<int> arr, int low, int high, SortType type) {
    if (low >= high) return arr;

    final middle = (low + high) ~/ 2;

    mergeSort(arr, low, middle, type);
    mergeSort(arr, middle + 1, high, type);

    for (var i = low; i <= middle; i++) {
      workArr[i] = arr[i];
    }

    for (var i = middle + 1, j = high; i <= high; i++, j--) {
      workArr[i] = arr[j];
    }

    var i = low, j = high;
    for (var k = low; k <= high; k++) {
      if (compare(workArr[j], workArr[i], type)) {
        arr[k] = workArr[i++];
      } else {
        arr[k] = workArr[j--];
      }
    }

    return arr;
  }
}
